modules/csg/csg_shape.cpp

```diff
@@ -32,6 +32,306 @@
 
 #include "core/math/geometry_2d.h"
 
+#include "thirdparty/manifold/src/manifold/include/manifold.h"
+#include "thirdparty/manifold/src/utilities/include/public.h"
+
+// Static helper functions.
+
+struct CSGBrush {
+	struct Face {
+		Vector3 vertices[3];
+		Vector2 uvs[3];
+		AABB aabb;
+		bool smooth = false;
+		bool invert = false;
+		int material = 0;
+	};
+
+	Vector<Face> faces;
+	Vector<Ref<Material>> materials;
+
+	manifold::Manifold manifold;
+	enum {
+		MANIFOLD_PROPERTY_POS_X = 0,
+		MANIFOLD_PROPERTY_POS_Y,
+		MANIFOLD_PROPERTY_POS_Z,
+		MANIFOLD_PROPERTY_UV_X,
+		MANIFOLD_PROPERTY_UV_Y,
+		MANIFOLD_PROPERTY_MATERIAL,
+		MANIFOLD_PROPERTY_SMOOTH,
+		MANIFOLD_PROPERTY_INVERT,
+		MANIFOLD_MAX
+	};
+	static constexpr int MANIFOLD_TRIANGLE = 3;
+
+	inline void _regen_face_aabbs() {
+		for (int i = 0; i < faces.size(); i++) {
+			faces.write[i].aabb = AABB();
+			faces.write[i].aabb.position = faces[i].vertices[0];
+			faces.write[i].aabb.expand_to(faces[i].vertices[1]);
+			faces.write[i].aabb.expand_to(faces[i].vertices[2]);
+		}
+	}
+
+	void pack_manifold(const float p_snap, HashMap<uint32_t, Ref<Material>> &r_materials) {
+		if (faces.is_empty()) {
+			return;
+		}
+		faces.sort_custom<CSGBrush::MaterialIndexComparator>();
+
+		HashSet<int32_t> unique_materials;
+		for (CSGBrush::Face &face : faces) {
+			unique_materials.insert(face.material);
+		}
+
+		LocalVector<CSGBrush> split_brushes;
+		CowData<CSGBrush::Face>::Size unique_face_i = 0;
+		for (int32_t material_id : unique_materials) {
+			uint32_t reserved_id = manifold.ReserveIDs(1);
+			CSGBrush current_brush;
+			Ref<Material> material;
+			if (material_id >= 0 && material_id < materials.size()) {
+				material = materials[material_id];
+			}
+			CSGBrush::Face starting_face = faces[unique_face_i];
+			while (unique_face_i < faces.size()) {
+				CSGBrush::Face face = faces[unique_face_i];
+				if (starting_face.material != face.material) {
+					break;
+				}
+				face.material = reserved_id;
+				current_brush.faces.push_back(face);
+				unique_face_i++;
+			}
+			manifold::MeshGL mesh;
```

Comment:
The problem is here: you're creating a mesh per material, but what you want is a single mesh where each material becomes a new run of triangles. Basically you need to be concatenating what you're doing here instead of starting over. Then perform the `Merge` outside of your materials loop.

Reply:
I'm going to try to take a look (chatted with @fire, I think it's mainly a matter of not having time to dig into this). I poked around and I believe the tutorial/example is [bindings/wasm/examples/three.ts](https://github.com/elalish/manifold/blob/c01f5df49a9fd5a766cfde55cba5412dc7d77f5e/bindings/wasm/examples/three.ts#L118-L144), but let me know if there's something more involved you both were using. I think I broadly understand the suggestion, though.

Reply:
Great to hear, and yes, you're correct. Don't hesitate to reach out if you have questions.

Reply:
I think I got it working: https://github.com/31/godot/commit/df33cbff3b4a68941debc1c9b2c8fe9b3ba7d97d. I don't have a great set of test cases to go through, but it seems to work and keep track of materials. I rewrote a chunk of the pack/unpack functionality while I was trying to understand it better, apologies @fire for making the diff a little noisy.